### ISSUE TITLE
Model will not train.

### ISSUE THREAD

**Issue by matrix-atom** [Fediory/HVI-CIDNet#51]
I am attempting to trained a new model from scratch for the SID dataset (I have structured the dataset as described), but the model will not even begin the first epoch. The necessary CUDA drivers and version is installed (pip install torch==1.13.1+cu117 torchvision==0.14.1+cu117 torchaudio==0.13.1 --extra-index-url https://download.pytorch.org/whl/cu117).

Attached is a screenshot of my issue. 

![Image](https://github.com/user-attachments/assets/ff527430-3cb3-423d-948a-fa2f963fa603)

**Comment by matrix-atom** [comment]
![Image](https://github.com/user-attachments/assets/64e91e12-0d25-47a5-b321-46daa4f2eab7)

**Comment by Fediory** [comment]
Don't you have this issue when training on other datasets?

**Comment by matrix-atom** [comment]
I haven't tried other datasets. I only have the SID data and that's the only one I am currently interested in.

**Comment by matrix-atom** [comment]
I thought this was happening due to a CUDA discrepancy but I cannot track it down. I don't see any obvious problem with CUDA.

Below is my nvidia driver. I did downgrade previously but that did not help either.

![Image](https://github.com/user-attachments/assets/a5d71ee6-4ae2-4431-ad4b-da84a70974ea)

**Comment by Fediory** [comment]
My suggestion would be to try the LOL dataset first to see if it has the same problem. If the phenomenon is consistent, then the CUDA version may be too high and further analysis is required. From the current state of the screenshots, I don't see any issues, just a bit too low GPU usage.

**Comment by matrix-atom** [comment]
Thank you for the suggestion; the LOL dataset is currently training.

What do you believe may have caused the SID dataset training issue? I would think that I have plenty of RAM at 32 GB to handle that data with the dataloader.

**Comment by Fediory** [comment]
Maybe you can lower the threads of dataloader in `data/options.py` line 14.

**Comment by matrix-atom** [comment]
Threads default=1 still shows the same problem.

**Comment by matrix-atom** [comment]
Just pulled the repo again, moved the SID data into datasets folder, and still have the same problem with only a single thread. Odd that LOL worked just fine but not SID. 

Any other suggestions?

**Comment by Fediory** [comment]
I think it's the program going into a dead loop in one of the files. I don't have enough information right now, so debug and locate it yourself to see the exact code that is stuck.

**Comment by matrix-atom** [comment]
Yeah I have tried that already. Not sure why but the SID data gets stuck iterating over the batch variable. Note, there is no error when loading the SID data using dataloader. Below is the process of debugging:

1. Step into the for loop using tqdm in train.py [line 47, "for batch in tqdm(training_data_loader):"]
2. Continue to the for the for loop over **iterable** in tqdm\std.py (line 1178). 
3. Step into the for loop to yield batch **obj** 

But it won't yield the **obj** batch variable for the SID dataset.

![Image](https://github.com/user-attachments/assets/39269afd-1bdd-4718-b53b-e08ee08fc0ae)

When using LOL dataset, I can see the batch **obj** variable shown below:

![Image](https://github.com/user-attachments/assets/ce99de99-6ba9-487e-bde7-7ae67621961f)

**Comment by matrix-atom** [comment]
I have been tweaking the options quite a bit; larger crop size up to 1024x1024, lower batch number, 1 thread, etc. 

Do you have the options or any other information about how you performed training with the SID data? I would like to just try to replicate that training, so if you did perform training with SID than I can just repeat your experiment.

**Comment by Fediory** [comment]
We actually made some changes to the SID dataset. Make sure you download Sony-Total-Dark from our Readme.

**Comment by matrix-atom** [comment]
I have been using your SID dataset. I wrote a script to set up the training data into the appropriate directories (\test, \train; \long, \short).

So, that is not the problem.

**Comment by matrix-atom** [comment]
I have previously downloaded your Sony_total_dark.zip dataset and wrote a script to arrange files into the appropriate directories.

After downloading the data again just a bit ago, training works! Thank you for that recommendation.

Was this data recently updated?

I see that it is in RGB format with size M/2 x N/2 from the original Bayer pattern image size (or native .arw file size). What did you use to convert it to RGB? 

Just curious because I have used the MATLAB raw2planar function when training their Bayer pattern low-light enhancement network (https://www.mathworks.com/help/images/learning-to-see-in-the-dark.html) as well as opencv's conversion in python (cv.COLOR_BGR2RGB) with that latter producing poor coloring to RGB.

**Comment by Fediory** [comment]
We did not update our SID dataset. It is possible that you have changed some folder names before. Because our dataset build searches based on the folder, if we can't find the corresponding pairing data, we will keep searching in the while loop.

We convert RAW to sRGB images because our method is based on the sRGB enhancement method. However, because this conversion process will definitely result in information loss, no matter what method is used to convert, that is why the enhanced image is poor.

**Comment by matrix-atom** [comment]
> We convert RAW to sRGB images because our method is based on the sRGB enhancement method. 
>... conversion process will definitely result in information loss, no matter what method is used to convert, that is why the enhanced image is poor.

I read the paper, so I am aware that you built this network to work with RGB. As I stated in my last comment, I have already used another network that applies Bayer-to-RGB conversion for the truth data (MATLAB's raw2planar). That provided a quality conversion.

I am asking what conversion you specifically used to obtain your Bayer-to-sRGB data. I would like to know for testing purposes with challenging sensors like that of the Bayer pattern in the Sony data.

**Comment by matrix-atom** [comment]
Lastly, do you have any intention of updating the code to work with newer CUDA version?

I may do that myself so that I can train models on other machines that are incompatible with CUDA 11.7.

**Comment by Fediory** [comment]
> Lastly, do you have any intention of updating the code to work with more recent CUDA versions?
> 
> I may do that myself so that I can train models on other machines that are incompatible with CUDA 11.7.

We have tested the current code to be compatible with the latest cuda version, the only thing that may need to be changed is some version numbers in `requirement.txt`. However, we still keep the current version of the code because this is the version of the code we used in our paper to train to the optimal result.

**Comment by Fediory** [comment]
We use `rawpy` library to convert ARW to png. the code is as follows：
`raw.postprocess(use_camera_wb=True, half_size=True, no_auto_bright=True, output_bps=8)`

**Comment by matrix-atom** [comment]
> We use `rawpy` library to convert ARW to png. the code is as follows： `raw.postprocess(use_camera_wb=True, half_size=True, no_auto_bright=True, output_bps=8)`

Thank you! I will try this library.

**Comment by matrix-atom** [comment]
I have some results.

Here is a result from a model trained for 1000 epochs at batch size=4, learning rate=1e-4, crops size=512.

![Image](https://github.com/user-attachments/assets/601002e3-358a-430c-b4dc-7552bb715fd8)

Here is a result from a model trained for 1000 epochs at batch size=8, learning rate=1e-3, crops size=256.

![Image](https://github.com/user-attachments/assets/1dc3e5d1-67bc-475f-a636-b97091b53b53)

I will continue empirically altering hyperparameters. Any advice?
